service/http.proxy/src/main/java/org/kaazing/gateway/service/http/proxy/HttpProxyServiceHandler.java

```diff
@@ -94,10 +95,37 @@ public void initializeSession(IoSession session, ConnectFuture future) {
             HttpConnectSession connectSession = (HttpConnectSession) session;
             connectSession.setVersion(acceptSession.getVersion());
             connectSession.setMethod(acceptSession.getMethod());
-            connectSession.setRequestURI(acceptSession.getRequestURI());
+            String acceptPathInfo = acceptSession.getPathInfo().toString();
+            String connectPath = connectSession.getRequestURI().getPath();
+            String computedPath = composePath(acceptPathInfo, connectPath);
+            if ( !validateComputedPath(computedPath, connectPath) ) {
+                acceptSession.setStatus(CLIENT_NOT_FOUND);
+                acceptSession.close(false);
+                return;
+            }
+            connectSession.setRequestURI(URI.create(computedPath) );
             processRequestHeaders(acceptSession, connectSession);
         }
 
+        private String composePath(String acceptPath, String connectPath) {
+            boolean connectPathEndsInSlash = ( connectPath.lastIndexOf("/") == (connectPath.length()-1) ) ? true : false;
+            boolean acceptPathStartsWithSlash = acceptPath.indexOf("/") == 0 ? true : false;
+            if (connectPathEndsInSlash && acceptPathStartsWithSlash) {
+                acceptPath = acceptPath.substring(1);
+            }
+            return connectPath + acceptPath;
+        }
+
+        private boolean validateComputedPath(String path, String connectPath) {
+
+            path = URI.create(path).normalize().toString();
+            if (path.indexOf("/\\.\\./") == 0 || path.indexOf(connectPath) != 0) {
```

Comment:
I think we can reduce the logic. I think validation should include only request path and accept path. Then if validation passes, just use the path to resolve connect URI (I think you can use resolve() method)